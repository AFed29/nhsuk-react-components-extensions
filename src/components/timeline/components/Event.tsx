import React, { ReactNode } from 'react';

export interface EventProps {
  title: ReactNode;
  instigator: ReactNode;
  date: Date;
  description?: Array<ReactNode>;
  action?: ReactNode;
}

const dateOptions = {
  day: 'numeric',
  month: 'short',
  year: 'numeric',
  hour: 'numeric',
  minute: 'numeric',
  second: 'numeric',
  hour12: true,
};

const Event: React.FC<EventProps> = ({ title, instigator, date, description = [], action }) => (
  <div className="nhsuk-timeline__event nhsuk-u-margin-bottom-2">
    <h2 className="nhsuk-timeline__title">
      <span className="nhsuk-timeline__status">{title}</span>
      <span className="nhsuk-timeline__by"> by {instigator}</span>
      {action}
    </h2>
    <p className="nhsuk-timeline__date">
      <time dateTime={date.toString()}>{date.toLocaleString('en-GB', dateOptions)}</time>
    </p>
    {description.map((descriptionItem, index) => (
      <p key={index.toString()} className="nhsuk-timeline__description-item">
        {descriptionItem}
      </p>
    ))}
  </div>
);

export default Event;
