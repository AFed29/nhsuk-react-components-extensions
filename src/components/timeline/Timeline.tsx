import React from 'react';
import Event, { EventProps } from './components/Event';

type TimelineProps = {
  events: Array<EventProps>;
};

const Timeline: React.FC<TimelineProps> = ({ events }) => (
  <div className="nhsuk-timeline">
    {events.map(({ title, instigator, date, description }, index) => (
      <Event
        key={index.toString()}
        title={title}
        instigator={instigator}
        date={date}
        description={description}
      />
    ))}
  </div>
);

export default Timeline;
