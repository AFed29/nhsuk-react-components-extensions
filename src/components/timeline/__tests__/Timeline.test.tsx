import React from 'react';
import { shallow } from 'enzyme';
import Timeline from '../Timeline';
import Event from '../components/Event';

const testEvents = [
  {
    title: 'Result sent',
    instigator: 'System',
    date: new Date('2019-11-19'),
    description: [
      'Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised',
      'Test Date: 19-Oct-2020, 9:00:00 am',
    ],
  },
  {
    title: 'Patient deferred',
    instigator: 'James Smith',
    date: new Date('2019-11-19'),
    description: [
      'Defer Reason: Pregnancy',
      'CRM Case Number: CAS-12345-ABCDE',
      'Comments: Requested via the GP form',
    ],
  },
  {
    title: 'Next test due date changed',
    instigator: 'System',
    date: new Date('2019-11-19'),
    description: ['Next test due date changed to 10-Oct-2021'],
  },
  {
    title: 'Test result added',
    instigator: 'System (Edifax)',
    date: new Date('2019-11-19'),
    description: ['Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised'],
  },
];

describe('Timeline', () => {
  it('matches snapshot', () => {
    const timeline = shallow(<Timeline events={testEvents} />);

    expect(timeline).toMatchSnapshot();

    timeline.unmount();
  });

  it('should have correct classes applied', () => {
    const timeline = shallow(<Timeline events={testEvents} />);

    expect(timeline.hasClass('nhsuk-timeline')).toBeTruthy();

    timeline.unmount();
  });

  describe('Event', () => {
    it('matches snapshot', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event).toMatchSnapshot();

      event.unmount();
    });

    it('should have correct classes applied', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.hasClass('nhsuk-timeline__event')).toBeTruthy();

      event.unmount();
    });

    it('should have correct title', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__title').text()).toBe('Result sent by System');

      event.unmount();
    });

    it('should have correct instigator', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__by').text()).toBe(` by ${testEvents[0].instigator}`);

      event.unmount();
    });

    it('should have correct date', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      const dateOptions = {
        day: 'numeric',
        month: 'short',
        year: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
        second: 'numeric',
        hour12: true,
      };

      expect(event.find('.nhsuk-timeline__date time').text()).toBe(
        testEvents[0].date.toLocaleString('en-GB', dateOptions),
      );

      event.unmount();
    });
  });
});
