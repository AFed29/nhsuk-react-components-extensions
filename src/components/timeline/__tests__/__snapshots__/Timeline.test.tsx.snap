// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Event matches snapshot 1`] = `
<div
  className="nhsuk-timeline__event nhsuk-u-margin-bottom-2"
>
  <h2
    className="nhsuk-timeline__title"
  >
    Result sent
    <span
      className="nhsuk-timeline__by"
    >
       by 
      System
    </span>
  </h2>
  <p
    className="nhsuk-timeline__date"
  >
    <time
      dateTime="Tue Nov 19 2019 00:00:00 GMT+0000 (Greenwich Mean Time)"
    >
      19 Nov 2019, 0:00:00 am
    </time>
  </p>
  <p
    className="nhsuk-timeline__description-item"
    key="0"
  >
    Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised
  </p>
  <p
    className="nhsuk-timeline__description-item"
    key="1"
  >
    Test Date: 19-Oct-2020, 9:00:00 am
  </p>
</div>
`;

exports[`Timeline matches snapshot 1`] = `
<div
  className="nhsuk-timeline"
>
  <Event
    date={2019-11-19T00:00:00.000Z}
    description={
      Array [
        "Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised",
        "Test Date: 19-Oct-2020, 9:00:00 am",
      ]
    }
    instigator="System"
    key="0"
    title="Result sent"
  />
  <Event
    date={2019-11-19T00:00:00.000Z}
    description={
      Array [
        "Defer Reason: Pregnancy",
        "CRM Case Number: CAS-12345-ABCDE",
        "Comments: Requested via the GP form",
      ]
    }
    instigator="James Smith"
    key="1"
    title="Patient deferred"
  />
  <Event
    date={2019-11-19T00:00:00.000Z}
    description={
      Array [
        "Next test due date changed to 10-Oct-2021",
      ]
    }
    instigator="System"
    key="2"
    title="Next test due date changed"
  />
  <Event
    date={2019-11-19T00:00:00.000Z}
    description={
      Array [
        "Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised",
      ]
    }
    instigator="System (Edifax)"
    key="3"
    title="Test result added"
  />
</div>
`;
